### PR TITLE
Bug fix + tests: keras.backend.batch_dot

### PR DESCRIPTION
**Pull request by farizrahman4u**

### CODE REVIEW

File: tensorflow/python/keras/backend.py [hunk 1]
```diff
@@ -1506,49 +1514,106 @@ def batch_dot(x, y, axes=None):
...
   ```
   """
+  x_shape = int_shape(x)
+  y_shape = int_shape(y)
+
+  x_ndim = len(x_shape)
+  y_ndim = len(y_shape)
+
+  if x_ndim < 2 or y_ndim < 2:
+    raise ValueError('Can not do batch_dot on inputs '
```
Comment (caisq) [inline tensorflow/python/keras/backend.py]: nit: 'Can not' --> 'Cannot'

File: tensorflow/python/keras/backend.py [hunk 2]
```diff
@@ -1506,49 +1514,106 @@ def batch_dot(x, y, axes=None):
...
+                      'Received inputs with shapes ' +
+                      str(x_shape) + ' and ' +
+                      str(y_shape) + '.')
+
+  x_batch_size = x_shape[0]
+  y_batch_size = y_shape[0]
+
+  if x_batch_size is not None and y_batch_size is not None:
+    if x_batch_size != y_batch_size:
+        raise ValueError('Can not do batch_dot on inputs '
```
Comment (caisq) [inline tensorflow/python/keras/backend.py]: nit: 'Can not' --> 'Cannot'

File: tensorflow/python/keras/backend.py [hunk 3]
```diff
@@ -65,6 +65,7 @@
 
 py_all = all
 py_sum = sum
+py_any = any
```
Comment (caisq) [inline tensorflow/python/keras/backend.py]: Can you modify the PR description to briefly describe what bug this change is fixing?

File: tensorflow/python/keras/backend.py [hunk 4]
```diff
@@ -1506,49 +1514,106 @@ def batch_dot(x, y, axes=None):
...
+                          'with different batch sizes. '
+                          'Received inputs with shapes ' +
+                          str(x_shape) + ' and ' +
+                          str(y_shape) + '.')
+
   if isinstance(axes, int):
-    axes = (axes, axes)
-  x_ndim = ndim(x)
-  y_ndim = ndim(y)
+    axes = [axes, axes]
```
Comment (caisq) [inline tensorflow/python/keras/backend.py]: Use tuple?

Comment (caisq) [reply tensorflow/python/keras/backend.py]: Same below.

File: tensorflow/python/keras/backend.py [hunk 5]
```diff
@@ -1506,49 +1514,106 @@ def batch_dot(x, y, axes=None):
...
-      idx = x_ndim - 1
-    out = array_ops.squeeze(out, list(range(idx, idx + diff)))
-  if ndim(out) == 1:
-    out = expand_dims(out, 1)
-  return out
+      axes = [x_ndim - 1, y_ndim - 2]
+
+  if py_any([isinstance(a, (list, tuple)) for a in axes]):
+    raise ValueError('Multiple target dimensions are not supported. ' +
+                     'Expected: None, int, (int, int), ' +
```
Comment (caisq) [inline tensorflow/python/keras/backend.py]: nit: 'None, int, or (int, int)'. Add 'or'.

File: tensorflow/python/keras/backend.py [hunk 6]
```diff
@@ -1506,49 +1514,106 @@ def batch_dot(x, y, axes=None):
...
+                     'Provided: ' + str(axes))
+
+  # if tuple, convert to list
```
Comment (caisq) [inline tensorflow/python/keras/backend.py]: nit: Follow Google Python style. Comments like this should start with a uppercase letter and end with a period.
https://github.com/google/styleguide/blob/gh-pages/pyguide.md

File: tensorflow/python/keras/backend.py [hunk 7]
```diff
@@ -1506,49 +1514,106 @@ def batch_dot(x, y, axes=None):
...
+
+  # convert negative indices
+  if axes[0] < 0:
+    axes[0] += x_ndim
+  if axes[1] < 0:
+    axes[1] += y_ndim
+
+  # sanity checks
+  if 0 in axes:
+    raise ValueError('Can not perform batch_dot over axis 0.'
```
Comment (caisq) [inline tensorflow/python/keras/backend.py]: A unit test is needed for this error condition.

File: tensorflow/python/keras/backend.py [hunk 8]
```diff
@@ -1506,49 +1514,106 @@ def batch_dot(x, y, axes=None):
...
+                      'If your inputs are not batched,'
+                      ' add a dummy batch dimension to your '
+                      'inputs using K.expand_dims(x, 0)')
+
+  a0, a1 = axes
+  d1 = x_shape[a0]
+  d2 = y_shape[a1]
+
+  if d1 is not None and d2 is not None and d1 != d2:
+    raise ValueError('Can not do batch_dot on inputs with shapes ' +
```
Comment (caisq) [inline tensorflow/python/keras/backend.py]: Need a unit test for this condition (if it doesn't already exist).

**Review by caisq — CHANGES_REQUESTED** [review]

File: tensorflow/python/keras/backend.py (hunk 3 above)
Comment (farizrahman4u) [reply tensorflow/python/keras/backend.py]: This is standard practice, as you can see in the lines above. `all`, `sum` and `any` backend functions overrides python's built in `all`, `sum` and `any`.

Comment (caisq) [reply tensorflow/python/keras/backend.py]: No, I was talking about this pull request in general. What bug is it trying to fix? If there is a GitHub issue, please link to it.

Comment (farizrahman4u) [reply tensorflow/python/keras/backend.py]: You can see that the tests (added in this PR) will fail without the changes to backend.py. These changes have already been made to keras-team/keras.

File: tensorflow/python/keras/backend.py [hunk 9]
```diff
@@ -1464,30 +1465,37 @@ def batch_dot(x, y, axes=None):
...
-  Arguments:
+  # Arguments
       x: Keras tensor or variable with `ndim >= 2`.
       y: Keras tensor or variable with `ndim >= 2`.
-      axes: list of (or single) int with target dimensions.
-          The lengths of `axes[0]` and `axes[1]` should be the same.
+      axes: int or tupe(int, int). Target dimensions to be reduced.
 
-  Returns:
+  # Returns
```
Comment (fchollet) [inline tensorflow/python/keras/backend.py]: Please fix incorrect docstring formatting (see other docstrings in the TF codebase)

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: tensorflow/python/keras/backend.py (hunk 4 above)
Comment (farizrahman4u) [reply tensorflow/python/keras/backend.py]: Has to be mutable for in-place subtraction.

**Comment by farizrahman4u** [comment]
@fchollet Are unit tests really needed for error conditions?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ymodak: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

File: tensorflow/python/keras/backend.py (hunk 3 above)
Comment (caisq) [reply tensorflow/python/keras/backend.py]: I'm aware of the current bugs in Dot. I just want to make sure the commit message documents the change better.

File: tensorflow/python/keras/backend.py (hunk 4 above)
Comment (caisq) [reply tensorflow/python/keras/backend.py]: Ack. Thanks.

File: tensorflow/python/keras/backend.py (hunk 9 above)
Comment (caisq) [reply tensorflow/python/keras/backend.py]: Ping @farizrahman4u Can you fix this so we can merge the PR? I'll leave it up to you if you want to add unit test for error conditions. It's not critical, but nice to have.

**Comment by farizrahman4u** [comment]
There were more changes in batch_dot at keras-team/keras. PR has been updated with those changes.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @caisq, @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @caisq, @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @caisq, @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 44 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @caisq, @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 59 days with no activity and the `awaiting review` label has been applied.
